afterburner/src/main/java/com/fasterxml/jackson/module/afterburner/util/MyClassLoader.java

```diff
@@ -56,52 +58,141 @@ public static boolean canAddClassInPackageOf(Class<?> cls)
             }
+            // important: must also resolve the newly-created class.
+            resolveClass(newClass);
+            return newClass;
         }
+    }
 
-        // but if that doesn't fly, try to do it from our own class loader
-    
-        try {
-            impl = defineClass(className.getDottedName(), byteCode, 0, byteCode.length);
-        } catch (LinkageError e) {
-            Throwable t = e;
-            while (t.getCause() != null) {
-                t = t.getCause();
+    /**
+     * Attempt to load (and resolve) the class using the parent class loader (if it is configured and present).
+     * This method will return {@code null} if the parent classloader is not configured or cannot be retrieved.
+     *
+     * @param className Interface or abstract class that class to load should extend or implement
+     * @param byteCode  the generated bytecode for the class to load
+     * @return          the loaded class, or {@code null} if the class could not be loaded on the parent classloader.
+     */
+    private Class<?> loadAndResolveUsingParentClassloader(ClassName className, byte[] byteCode)
+    {
+        ClassLoader parentClassLoader;
+        if (!_cfgUseParentLoader || (parentClassLoader = getParent()) == null) {
+            return null;
+        }
+        synchronized (getParentClassLoadingLock(parentClassLoader, className.getDottedName())) {
+            // First: check to see if the parent classloader has loaded it already
+            Class<?> impl = findLoadedClassOnParent(parentClassLoader, className.getDottedName());
+            if (impl != null) {
+                return impl;
             }
-            throw new IllegalArgumentException("Failed to load class '"+className+"': "+t.getMessage(), t);
+
+            // Important: bytecode is generated with a template name (since bytecode itself
+            // is used for checksum calculation) -- must be replaced now, however
+            replaceName(byteCode, className.getSlashedTemplate(), className.getSlashedName());
+
+            // Second: define a new class instance on the parent classloder using the bytecode
+            impl = defineClassOnParent(parentClassLoader, className.getDottedName(), byteCode, 0, byteCode.length);
+            // important: must also resolve the newly-created class.
+            resolveClassOnParent(parentClassLoader, impl);
+            return impl;
+        }
+    }
+
+    /**
+     * Get the class loading lock from the parent class loader for loading the named class.
+     *
+     * @param parentClassLoader     The parent ClassLoader
+     * @param className             The name of the to-be-loaded class
+     * @throws IllegalStateException if there is no parent ClassLoader or the lock object could not be retrieved
+     */
+    private Object getParentClassLoadingLock(ClassLoader parentClassLoader, String className)
```

Comment:
Is there particular reason it must be something internal to JDK, to lock on? Access to that object is potentially quite expensive, and with later JDK versions likely very fragile.

Reply:
I think it needs to use the parent `ClassLoader`'s internal lock for complete correctness. `ClassLoader#loadClass` acquires the lock, then (potentially) calls `ClassLoader#findClass`, which may ultimately call `ClassLoader#defineClass`. `#defineClass` will fail if it is called twice for the same class name (which is guarded against by the synchronization in `#loadClass`). Acquiring the same lock used by `ClassLoader#loadClass` guarantees that no other mechanism (outside of Afterburner) could load a class with the name we are using between the time we check to see whether it is defined and attempt to define it.

In our case, though, if we assume that `MyClassLoader` will be the only class loader implementation than attempts to load/define a class with a specific name (a name which always includes the 'Creator4JacksonDeserializer' suffix and the bytecode hash, so this seems like a safe assumption), then we should be safe in relaxing this requirement and just synchronizing on our own static lock.

Reply:
I'd feel better about our own lock unless we can think of specific case where someone or something might try to load class with same name concurrently; at least outside of Afterburner.
If the concern is for concurrent Afterburner instances from same JVM, probably would be ok to use static singleton lock (... yeah, not a big fan of those things, who is, but here it seems inevitable...). And I guess that is a valid concern, so even though ideally one tries to avoid multiple instances of `ObjectMapper` -- and esp. so with Afterburner generating classes on the fly! -- there are some legit cases when 2 or 3 mappers are not for different configurations.

So, if this could be done that is great.

Also: just so I won't forget to follow up, I added this on:

https://github.com/FasterXML/jackson-future-ideas/wiki/Jackson-Work-in-Progress

as an urgent issue.